datafusion/execution/src/parquet.rs

```diff
@@ -0,0 +1,62 @@
+use arrow::datatypes::SchemaRef;
+use dashmap::DashMap;
+use datafusion_common::config::TableParquetOptions;
+use datafusion_common::DataFusionError;
+#[cfg(feature = "parquet")]
+use parquet::encryption::decrypt::FileDecryptionProperties;
+#[cfg(feature = "parquet")]
+use parquet::encryption::encrypt::FileEncryptionProperties;
+use std::sync::Arc;
+
+/// Trait for types that generate file encryption and decryption properties to
+/// write and read encrypted Parquet files.
+/// This allows flexibility in how encryption keys are managed, for example, to
+/// integrate with a user's key management service (KMS).
+#[cfg(feature = "parquet")]
+pub trait EncryptionFactory: Send + Sync + std::fmt::Debug + 'static {
+    /// Generate file encryption properties to use when writing a Parquet file.
+    fn get_file_encryption_properties(
+        &self,
+        options: &TableParquetOptions,
+        schema: &SchemaRef,
+        file_path: &str,
```

Comment:
I think it would probably be better to use std::Path here for portability.

Reply:
One thing I wasn't sure about is that just using a Path loses the type of object store being used, but there doesn't seem to be a common type used between writing and reading that holds both these bits of information, and I think just using `Path` is fine.